File: 2021/python/day01.py
def Run(filepath):
    with open(filepath) as input:
        measurements = list(map(int, input.read().split('\n')))

    print('Day 1')
    Part1(measurements)
    Part2(measurements)

def Part1(measurements):
    increases = 0
    previousDepth = None

    for depth in measurements:
        #Increment if this isn't the first measurement and the depth has increased
        if previousDepth != None and depth > previousDepth:
            increases += 1

        previousDepth = depth

    print('Part 1: ' + str(increases))
    #The result should be 1791

def Part2(measurements):
    increases = 0
    previousSum = None

    #Run as long as there are complete sets of 3 measurements to add together
    for index in range(len(measurements)-2):
        measurement1 = measurements[index]
        measurement2 = measurements[index+1]
        measurement3 = measurements[index+2]
        currentSum = measurement1 + measurement2 + measurement3
        
        #Increment if this isn't the first set of measurements and the sum has increased
        if previousSum != None and currentSum > previousSum:
            increases += 1

        previousSum = currentSum

    print('Part 2: ' + str(increases))
    #The result should be 1822

File: 2021/csharp/app/Day01.cs
namespace AdventOfCode;
public class Day01 : Day{
    public override Int64 Part1(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> readings = new List<string>(file);
        int increases = CountDepthIncreases(readings, 1);

        System.Console.WriteLine($"Day 1-1: {increases}");
        return increases;
    }
    public override Int64 Part2(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> readings = new List<string>(file);
        int increases = CountDepthIncreases(readings, 3);
        
        System.Console.WriteLine($"Day 1-2: {increases}");
        return increases;
    }
    static int CountDepthIncreases(List<string> readings, int groupSize){
        int currentSum = 0, previousSum = 0, increases = 0, depth, index = 0, indexDiff = groupSize - 1;

        //This loop uses indexDiff to run until there are no more full-sized groups to
        /// add together, e.g. a group of 3 items will start 2 indexes before the end 
        while (index < readings.Count - indexDiff){
            int i = 0;
            
            do{ ///Add at least the first depth reading to the sum, then loop if more depths exist
                depth = Convert.ToInt32(readings[index + i]);
                currentSum += depth;
                i++;
            } while (i < groupSize);

            //Increment if this isn't the first measurement and the depth has increased
            if (previousSum != 0 && currentSum > previousSum) {increases++;}
            previousSum = currentSum;
            currentSum = 0;
            index++;
        }
        return increases;
    }
}